### DISCUSSION TITLE
CI scripts are hard to hack on and hard to verify

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1096] [General]
We use a `test.sh` script in each crate. These scripts are all slightly different. The logic and test coverage are not always immediately obvious (e.g. why things are ordered the way they are). For example, trouble shooting a script can lead to changes that are not obvious to a developer later looking at the various scripts (even for the same developer (*cough* me)) especially when working on multiple crates at once. Further exasperating this, the script interacts with CI in non-obvious ways depending on how its called (e.g. if each env var is turned on for a separate script run or multiple env vars are turned on for a single script run).

All these problems are going to be compounded when we start crate smashing.

The obvious idea of unifying all the scripts does not work because the crates have different testing requirements.

I'm not sure if this problem stems from
- Bash is just plain obscure sometimes?
- Perhaps I have changed CI so much over the last few months that I'm misusing how the script was intended to be run?
- Perhaps I have changed the scripts so much over the last few months that I am misusing CI when running the script?

Does anyone have any ideas how to improve the situation please?

One idea I have is to  write a single script that does all the 'typical' stuff, put a copy in each crate and source it in the crates `test.sh`, include a comment saying "do not edit this script, if you want to make changes do so in `rust-bitcoin` so the changes can be propagated to all crates."

**Comment by dpc** [comment]
Use nix flakes, write your CI scripts in Rust, make it cached so it goes fast,..., profit!

**Reply by apoelstra** [reply]
I'd be down to use Rust rather than Python of course :) and wouldn't mind finally getting around to learning rix. But I wouldn't be any help setting this up.

**Comment by apoelstra** [comment]
Bitcoin has a separate repo with testing stuff, and it's mostly in Python https://github.com/bitcoin-core/bitcoin-maintainer-tools/ ... although their actual CI stuff is specific to bitcoin core (and libsecp, and whatever else is in that org..) and is even more complicated than our setup.

**Comment by Kixunil** [comment]
> All these problems are going to be compounded when we start crate smashing.

They shouldn't since we keep a single repository, so script can live in it.

> The obvious idea of unifying all the scripts does not work because the crates have different testing requirements.

Wow, are they really *that* different? If yes, it looks like architectural issue. We should figure out how to unify them.

> Bash is just plain obscure sometimes?

Yeah, Bash is a terrible language but launching commands from other langs tends to be annoying. Still may be worth it to change.

> put a copy in each crate

I'd really like if we could avoid making copies. If we could just download it directly from somewhere, check SHA256 and run it'd be awesome.

I also have a suspicion that our CI is not documented enough.

**Reply by tcharding** [reply]
> I also have a suspicion that our CI is not documented enough.

This might be the key. I'll try to document heavily as I make further changes.

**Comment by dpc** [comment]
BTW @tcharding . I've been recently working on something relevant: [a whole software shop based on Rust+Nix flakes combo](https://github.com/rustshop/rustshop). It is focused mostly on AWS/k8s stack, but under the hood it has to has to take care of all the usuals: building, CI/CD, lints etc for Rust + Nix. Some of the bits might be copy&paste reusable.

So happens that I was going to add some Github Action based workflows anyway, so I rushed it a bit and pushed [the first workflow](https://github.com/rustshop/rustshop/blob/61aa66dcfea88d21a760c91229b85dc09a9ba525/.github/workflows/ci-services.yml#L1). I plan to add `nixpkgs-fmt` and `rustfmt` lints today evening PST. If you want to see anything else setup for evaluation purposes I can prioritize as well.

You can also look at example of [including custom-build tooling in Rust](https://github.com/rustshop/rustshop/blob/61aa66dcfea88d21a760c91229b85dc09a9ba525/flake.nix#L157) (`rustshop` the binary is a special utility for managing/bootstrapping AWS orgs/accounts & k8s clusters among other stuff).

I just `nix develop` and get a shell with all the tools build and added to path - readily available to use, all the git hooks and similar automation installed and configured, etc. You can try it - just clone, `cd ...` and `nix develop`. You can try to add a commit, run the `rustshop --help`.

The goal of this project is have almost no code in anything other than Nix&Rust, and set it up in a really neat and optimized way. If you have any questions/opinions, feel free to post on Discussions there (I still need to write a good post discussing dev env & CI/CD)

**Reply by tcharding** [reply]
Sick, thanks man, I'll take a look. No need to prioritize, this is a ongoing itch I've wanted to scratch for a while but no rush.

**Reply by tcharding** [reply]
> Pretend you are delivering business value using Rust in your free time for no pay or benefits!

ROFL, epic!
